core/src/main/java/hudson/model/WorkspaceListener.java

```diff
@@ -3,25 +3,77 @@
 import hudson.ExtensionList;
 import hudson.ExtensionPoint;
 import hudson.FilePath;
+import hudson.Util;
+import hudson.console.ConsoleLogFilter;
+
+import java.io.OutputStream;
 
 public abstract class WorkspaceListener implements ExtensionPoint {
     
     /**
      * Called after a workspace is deleted successfully.
      * @param project
+     * @deprecated as of TODO_BEFORE_MERGE Use {@link #afterDelete(Job)}.
      */
     public void afterDelete(AbstractProject project) {
-        
+        if (Util.isOverridden(WorkspaceListener.class, getClass(), "afterDelete", Job.class)) {
+            // old client calling newer implementation. forward the call.
+            afterDelete((Job) project);
+        } else {
+            // happens only if the subtype fails to override neither beforeUse method
+            throw new AssertionError("The plugin '" + this.getClass().getName() + "' still uses " +
```

Comment:
This looks wrong to me.  The method was not abstract and now you force are forcing a client to override this or afterDelete(Job).

whereas you may have a listener that only implemented beforeUse in order to do something.